cpp/src/dataframe/device_table.cuh

```diff
@@ -0,0 +1,543 @@
+
+    for(gdf_size_type i = 0; i < _num_columns; ++i)
+    {
+      host_columns[i]->size = this->_num_rows;
+    }
+  }
+
+  gdf_size_type num_columns() const
+  {
+    return _num_columns;
+  }
+
+  __host__
+  gdf_column * get_column(gdf_size_type column_index) const
+  {
+    return host_columns[column_index];
+  }
+
+   __host__ 
+  gdf_column ** columns() const
+  {
+    return host_columns;
+  }
+
+  __host__ __device__
+  gdf_size_type num_rows() const
+  {
+    return _num_rows;
+  }
+
+  __device__ bool row_has_nulls(gdf_size_type row_index) const{
+      return not gdf_is_valid(d_row_valid, row_index);
+  }
+
+  /** 
+   * @brief  Gets the size in bytes of a row in the device_table, i.e., the sum of 
+   * the byte widths of all columns in the table
+   * 
+   * @returns The size in bytes of the row in the table
+   */
+  gdf_size_type get_row_size_bytes() const
+  {
+    return row_size_bytes;
+  }
+
+  /** 
+   * @brief  Packs the elements of a specified row into a contiguous byte-buffer
+   *
+   * This function is called by a single thread, and the thread will copy each element
+   * of the row into a single contiguous buffer. TODO: This could be done by multiple threads
+   * by passing in a cooperative group. 
+   * 
+   * @param index The row of the table to return
+   * @param row_byte_buffer A pointer to a preallocated buffer large enough to hold a 
+      row of the table 
+   * 
+   */
+  __device__
+  gdf_error pack_row(gdf_size_type row_index, void * row_byte_buffer) const
+  {
+    if(nullptr == row_byte_buffer) {
+      return GDF_DATASET_EMPTY;
+    }
+
+    // Pack the element from each column in the row into the buffer
+    for(gdf_size_type i = 0; i < _num_columns; ++i)
+    {
+      const gdf_dtype source_col_type = d_columns_types_ptr[i];
+
+      cudf::type_dispatcher(source_col_type,
+                            copy_element{},
+                            row_byte_buffer, i,
+                            d_columns_data_ptr[i], row_index);
+
+    }
+    return GDF_SUCCESS;
+  }
+
+  struct copy_element{
+    template <typename ColumnType>
+    __device__ __forceinline__
+    void operator()(void * target_column, gdf_size_type target_row_index,
+                    void const * source_column, gdf_size_type source_row_index)
+    {
+      ColumnType& target_value { static_cast<ColumnType*>(target_column)[target_row_index] };
+      ColumnType const& source_value{static_cast<ColumnType const*>(source_column)[source_row_index]};
+      target_value = source_value;
+    }
+
+  };
+
+  /**
+   * @brief  Copies a row from a source table to a target row in this table
+   *
+   * This device function should be called by a single thread and the thread
+   * will copy all of the elements in the row from one table to the other. 
+   * 
+   * TODO: In the future, this could be done by multiple threads by passing in a
+   * cooperative group.
+   * 
+   * FIXME: Does NOT set null bitmask for the target row.
```

Comment:
Yeah, I plan to fix it, but I wasn't sure how I wanted to do it yet.

One idea was to eliminate the need for this function by just refactoring the algorithm that uses `copy_row` to instead just use `gather`, and then I'll get the bitmask handling for free.

> (Copying the valid mask isn't that hard if you parallelize across rows (fast), or by using atomics (slow).)

 This function already assumes parallelizing across rows. However, I don't see how you can safely update the target's bitmask without using an `atomicOr`, since you're trying to update a single bit at a time.

Reply:
32 rows `__ballot_sync` and then update 32 bits at a time.

Reply:
The problem is that you aren't updating a contiguous set of rows in the output such that you can write 32 bits at a time. Writing the output is random access for the join/groupby use cases.

Reply:
Ahh, I see. I deal with this in apply_boolean_mask by staging in shared memory, and then writing coalesced to global memory. Not sure if that's possible here.